File: examples/octave/tlens_gaussian.m
% LightPipes for Octave Optical Toolbox

% load physical units and constants
physical  

% load some utility functions for watching propagation
propagate_

% pixel size on slm
DX=15*microns;
% wavelength
lambda=780*nm;

% lens applied directly on camera.
f_torus=40*cm;

% The radius of the center of the toroidal lens.
R0=128*DX;

R1 = .19*512*DX; % central radius of phase ring.
w1 = .23*512*DX; % width of phase ring.

%w = 0.12*40*mm/sqrt(2); % 40mm colimating lens
w = 1.71*mm/sqrt(2); % smaller beam

function phi = tlens(x,y,R0,f,lambda);
% function phi = tlens(x,y,R0,f,lambda);
%
%  This function calculates the phase to use for a
%   toroidal lens + quadratic lens.
    phi = 2*pi + rem(-pi*(sqrt(x.^2+y.^2) - R0).^2 / (f * lambda), 2*pi);
endfunction

function phi = phase_ring(x, y, R, w)
% function phi = phase_ring(x, y, R, w)
%
%  This function calculates the phase ring that is to be added to the phase of
%  the toroidal lens at the SLM.
    phi = 0 * x;
    r2 = (x.^2 + y.^2);
    Ri_2 = (R-0.5*w)^2.0;
    Rf_2 = (R+0.5*w)^2.0;
    phi(find( r2 >= Ri_2 & r2 < Rf_2 )) = pi;
endfunction

function plot_phase(phase, div)
% just a helper function to plot the phase and subsample first to get gnuplot
% to quit taking forever.
    gset view map
    gset nosurf
    gset nohidden3d
    gset pm3d at b

    gsplot phase(1:div:rows(phase),1:div:columns(phase));
endfunction


[xx yy] = meshgrid( x=(-255.5:1:255.5)*DX, y=x);
torus_phase = tlens (xx,yy,R0,f_torus,lambda) + phase_ring(xx,yy,R1,w1);

%plot_phase(torus_phase,4);
input('press enter to continue');



% propagate the field from the SLM and take a look
F1 = LPBegin(512*DX, lambda, 512);              % initial field
F2 = LPGaussAperture(w, 0, 0, 1,        F1);    % gaussian input
F3 = LPCircAperture(256*DX,0,0,         F2);    % Iris in front of SLM
F4 = F3; F4.F .*= exp(i * torus_phase );        % apply SLM phase
F5 = LPForvard( f_torus,                F4 );
%F5 = propagate(f_torus, f_torus/25,     F4, 0, 0, 4, 256);

plot_phase(F5.F .* conj(F5.F),4);

File: examples/python/tlens_gaussian.py
#!/usr/bin/env python
"""
LightPipes for Octave Optical Toolbox
"""

from common import Propagate
from lightpipes import Field, lg
from pylab import *

m=1
nm=1e-9*m
um=1e-6*m
mm=1e-3*m
cm=1e-2*m

N               = 512           # Number of pixels
grid_dx         = 15*um         # pixel size on slm
side_length     = N * grid_dx   # physical size of SLM
wavelength      = 780*nm        # Wavelength
f_torus         = 40*cm         # lens applied directly on camera
R0              = .25*N*grid_dx # radius of center of toroidal lens
R1              = .19*N*grid_dx # central radius of phase ring.
w1              = .23*N*grid_dx # width of phase ring.
#w               = 0.12*40*mm/sqrt(2) # 40mm colimating lens
w               = 1.71*mm/sqrt(2) # smaller beam

def tlens(x,y,R0,f,wavelength):
  """
  This function calculates the phase to use for a
  toroidal lens + quadratic lens.
  """
  return 2*pi + \
    (-pi*(sqrt(x**2+y**2) - R0)**2 / (f * wavelength) ) % (2*pi)


def phase_ring(x, y, R, w):
  """
  This function calculates the phase ring that is to be added to the phase of
  the toroidal lens at the SLM.
  """
  phi = zeros(x.shape)
  r2 = x**2 + y**2
  Ri_2 = (R - 0.5*w)**2
  Rf_2 = (R + 0.5*w)**2
  phi[ nonzero( r2 < Rf_2 ) ] = pi
  phi[ nonzero( r2 < Ri_2 ) ] = 0
  return phi


x = y = grid_dx * r_[-255.5:256:1.]
[xx, yy] = meshgrid(x,y)
torus_phase = tlens(xx,yy,R0,f_torus,wavelength) + phase_ring(xx,yy,R1,w1)

propagate = Propagate()

#print 'press enter to continue'
#propagate.imshow( torus_phase )
#raw_input()

# propagate the field from the SLM and take a look
F = Field( N, side_length, wavelength)     # initial field
F.gaussian_aperture(w)                     # gaussian input
F.circular_aperture(side_length/2.)        # Iris in front of SLM
F.value[:] *= exp(1j * torus_phase )       # apply SLM phase
F.forvard(f_torus)
propagate(F, z=f_torus, dz=f_torus/25.)

print 'press enter to show final phase...'
raw_input()
propagate.imshow( angle(F.value) )

print 'press enter to show final intensity...'
raw_input()
propagate.imshow( (F.value * F.value.conj()).real )

print 'press enter to exit'
raw_input()
